File: backend/utils/chunking.py
import re
from typing import List

def split_into_sections(raw_text: str, max_words_per_section: int = 400) -> List[str]:
    """
    Splits raw story text into coherent sections based on paragraphs and word limits.
    """
    cleaned = raw_text.strip().replace("\r\n", "\n")
    if not cleaned:
        return []

    # Split by double newline (paragraphs) first
    raw_paragraphs = [p.strip() for p in re.split(r"\n\s*\n", cleaned) if p.strip()]

    if not raw_paragraphs:
        return [cleaned]

    sections: List[str] = []
    current_section_words: List[str] = []

    for para in raw_paragraphs:
        para_words = [w for w in re.split(r"\s+", para) if w]

        # If adding this paragraph exceeds target word count and we already have content, flush
        if current_section_words and (len(current_section_words) + len(para_words) > max_words_per_section):
            sections.append(" ".join(current_section_words))
            current_section_words = list(para_words)
        else:
            current_section_words.extend(para_words)

    if current_section_words:
        sections.append(" ".join(current_section_words))

    # If story was one giant block without paragraph breaks, split by sentences
    if len(sections) == 1 and len(current_section_words) > max_words_per_section * 1.5:
        sentence_pattern = r"[^.!?।]+[.!?।]+|\S+"
        sentences = re.findall(sentence_pattern, cleaned) or [cleaned]
        sentence_sections: List[str] = []
        current_words: List[str] = []

        for sent in sentences:
            words = [w for w in re.split(r"\s+", sent.strip()) if w]
            if current_words and (len(current_words) + len(words) > max_words_per_section):
                sentence_sections.append(" ".join(current_words))
                current_words = list(words)
            else:
                current_words.extend(words)

        if current_words:
            sentence_sections.append(" ".join(current_words))

        return sentence_sections if sentence_sections else sections

    return sections if sections else [cleaned]

File: server/utils/chunking.ts
/**
 * Splits raw story into coherent sections (paragraphs/scenes)
 */

export function splitIntoSections(rawText: string, maxWordsPerSection: number = 400): string[] {
  const cleaned = rawText.trim().replace(/\r\n/g, '\n');
  if (!cleaned) return [];

  // Split by double newline (paragraphs) first
  const rawParagraphs = cleaned.split(/\n\s*\n/).map(p => p.trim()).filter(Boolean);

  if (rawParagraphs.length === 0) {
    return [cleaned];
  }

  const sections: string[] = [];
  let currentSectionWords: string[] = [];

  for (const para of rawParagraphs) {
    const paraWords = para.split(/\s+/).filter(Boolean);

    // If adding this paragraph exceeds target word count and we already have content, flush
    if (currentSectionWords.length > 0 && currentSectionWords.length + paraWords.length > maxWordsPerSection) {
      sections.push(currentSectionWords.join(' '));
      currentSectionWords = [...paraWords];
    } else {
      currentSectionWords.push(...paraWords);
    }
  }

  if (currentSectionWords.length > 0) {
    sections.push(currentSectionWords.join(' '));
  }

  // If story was one giant block without paragraph breaks, split by sentences
  if (sections.length === 1 && currentSectionWords.length > maxWordsPerSection * 1.5) {
    const sentences = cleaned.match(/[^.!?।]+[.!?।]+|\S+/g) || [cleaned];
    const sentenceSections: string[] = [];
    let currentWords: string[] = [];

    for (const sent of sentences) {
      const words = sent.trim().split(/\s+/).filter(Boolean);
      if (currentWords.length > 0 && currentWords.length + words.length > maxWordsPerSection) {
        sentenceSections.push(currentWords.join(' '));
        currentWords = [...words];
      } else {
        currentWords.push(...words);
      }
    }
    if (currentWords.length > 0) {
      sentenceSections.push(currentWords.join(' '));
    }
    return sentenceSections.length > 0 ? sentenceSections : sections;
  }

  return sections.length > 0 ? sections : [cleaned];
}
